service/frontend/nexus_handler.go

```diff
@@ -72,6 +73,7 @@ type nexusContext struct {
 	namespaceConcurrencyLimitInterceptor *interceptor.ConcurrentRequestLimitInterceptor
 	rateLimitInterceptor                 *interceptor.RateLimitInterceptor
 	responseHeaders                      map[string]string
+	responseHeadersMutex                 sync.RWMutex
```

Comment:
nit: you can use a `Mutex` here. We only read once before writing out the response.